### PR TITLE
Auth Added

### PR DESCRIPTION
**Pull request by tushar1210**
Authenticate and receive added

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,81 +1,53 @@
...
+import { type } from 'os';
+import { resolve } from 'path';
+import { rejects } from 'assert';
+import * as Promise from 'bluebird';
+let ws:WebSocket;
+var isConnected:boolean = false
+var isAuthenticated:boolean = false
+
+
+connect('123','qwerty')
```
Comment (yashanand1910) [inline src/index.ts]: Remove this test connect code. Add a test.ts file to test out the library, avoid adding these in the actual code

File: src/index.ts [hunk 2]
```diff
@@ -1,81 +1,61 @@
...
+
+//connect and authenticate
+function connect(moduleID:string,moduleSecret:string){
```
Comment (yashanand1910) [inline src/index.ts]: Even connect function itself should return a promise, that resolves on successful connect and auth and rejects on any error.

File: src/index.ts [hunk 3]
```diff
@@ -1,81 +1,61 @@
...
-			else
-				resolve();
-		});
-	});
+	
+	ws.on('open',()=>{
+		ws.send(JSON.stringify(authToken,null,'\t'))
+		
+	})
+	recieve().then((message)=>{
```
Comment (yashanand1910) [inline src/index.ts]: receive() should come before line#25

File: src/index.ts [hunk 4]
```diff
@@ -1,81 +1,61 @@
...
+// 	capabilityRequest.kaiId = kaiId
+// 	capabilityRequest.
+// }
+
+//Recieve : 
+
+function recieve(){
+	return new Promise((resolve,reject)=>{
+		ws.on('message',(data)=>{
+			var res:sdk.Response = JSON.parse(JSON.parse(JSON.stringify(data)))
```
Comment (yashanand1910) [inline src/index.ts]: Why are there 2 json.parse()?

**Review by yashanand1910 — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 4 above)
Comment (tushar1210) [reply src/index.ts]: Screenshot 2019-04-08 at 11 10 09 PM

File: src/index.ts (hunk 2 above)
Comment (tushar1210) [reply src/index.ts]: fixed

File: src/index.ts (hunk 1 above)
Comment (tushar1210) [reply src/index.ts]: fixed

File: src/index.ts (hunk 4 above)
Comment (rakshith-ravi) [reply src/index.ts]: Hold on a minute, so you're doing `JSON.parse(JSON.stringify(data))`. Wouldn't that just give you data directly? What's the point of the `JSON.stringify()` there?

Comment (tushar1210) [reply src/index.ts]: Screenshot 2019-04-09 at 9 07 20 AM

File: package.json [hunk 5]
```diff
@@ -17,7 +17,8 @@
   },
   "homepage": "https://github.com/vicara-hq/kai-nodejs#readme",
   "dependencies": {
-    "ws": "^6.1.0"
+    "bluebird": "^3.5.4",
```
Comment (rakshith-ravi) [inline package.json:20]: Why do we have bluebird again?

File: src/KaiEventEmitter.ts [hunk 6]
```diff
@@ -1,5 +1,16 @@
 import { EventEmitter } from "events";
 
 export default class KaiEventEmitter extends EventEmitter {
-	
+    ongesture(handler: any) {
```
Comment (rakshith-ravi) [inline src/KaiEventEmitter.ts]: `onGesture`, camelCased

File: src/KaiEventEmitter.ts [hunk 7]
```diff
@@ -1,5 +1,16 @@
...
+        this.addListener('gestureData',handler);
+    }
+
+    onauthentication(handler: any) {
```
Comment (rakshith-ravi) [inline src/KaiEventEmitter.ts]: `onAuthentication`

File: src/KaiEventEmitter.ts [hunk 8]
```diff
@@ -1,5 +1,16 @@
...
+        this.addListener('authentication', handler);
+    }
+
+    onconnect(handler: any) {
```
Comment (rakshith-ravi) [inline src/KaiEventEmitter.ts]: `onConnect`

**Review by rakshith-ravi — COMMENTED** [review]
Just needs some casing changes.
Also, please remove all `.DS_Store` files and add it to gitignore. Don't want that polluting the repo

File: src/index.ts (hunk 4 above)
Comment (rakshith-ravi) [reply src/index.ts]: `data` is an object. You wouldn't need to parse it, if I'm not wrong. `JSON.parse` takes a string only. That might be what's causing the issue

File: src/index.ts [hunk 9]
```diff
@@ -1,81 +1,155 @@
...
+    }else{
+        kaiEvents.emit(response.type,response.error)
+    }
 });
 
-ws.on('message', (data) => {
-	var input = JSON.parse(data.toString());
-	KaiEvents.emit('data', input);
+//TODO:-
+export function setCapabilities(kaiId:number|"default"|"defaultLeft"|"defaultRight",capabilitiesArr:string[]){
```
Comment (yashanand1910) [inline src/index.ts:54]: Why have you defined the **type** of the **kaiId** parameter over here, import all types from the **model** (here, _sdk.ts_). Import the type from _sdk.ts_ and use it.

File: src/index.ts [hunk 10]
```diff
@@ -1,81 +1,155 @@
...
-			resolve();
-		}
-		ws.send(json, err => {
-			if(err)
-				reject(err);
-			else
-				resolve();
-		});
-	});
+export function switchHand(kaiId:number|"default"|"defaultLeft"|"defaultRight"="default",hand:"left"|"right"="left"){
```
Comment (yashanand1910) [inline src/index.ts:119]: The type here for the hand parameter as well, import it from the model. Do not define types inside any other file or in this way, it is very inconsistent, always import it from the model.

File: src/index.ts [hunk 11]
```diff
@@ -1,81 +1,155 @@
...
+    let request:sdk.SwitchHandRequest={
```
Comment (yashanand1910) [inline src/index.ts:120]: This is the right way to use a type (sdk.SwitchHandRequest).

**Review by yashanand1910 — APPROVED** [review]
This can be merged as a first iteration. Does require improvements here and there for quality purposes but can be with time.

**Pull request merged**